File: flureadium/android/src/main/kotlin/dev/mulev/flureadium/FlutterPdfPreferences.kt
package dev.mulev.flureadium

import kotlinx.serialization.Serializable
import org.json.JSONObject
import org.readium.r2.navigator.preferences.Configurable
import org.readium.r2.navigator.preferences.Fit
import org.readium.r2.navigator.preferences.Spread

/**
 * PDF preferences for Flutter Readium.
 *
 * Maps to PDFPreferences in Flutter's reader_pdf_preferences.dart.
 */
@Serializable
data class FlutterPdfPreferences(
    val fit: FlutterPdfFit? = null,
    val scrollMode: FlutterPdfScrollMode? = null,
    val pageLayout: FlutterPdfPageLayout? = null,
    val offsetFirstPage: Boolean? = null,
) : Configurable.Preferences<FlutterPdfPreferences> {

    override fun plus(other: FlutterPdfPreferences): FlutterPdfPreferences =
        FlutterPdfPreferences(
            fit = other.fit ?: fit,
            scrollMode = other.scrollMode ?: scrollMode,
            pageLayout = other.pageLayout ?: pageLayout,
            offsetFirstPage = other.offsetFirstPage ?: offsetFirstPage
        )

    /**
     * Converts to Readium-compatible preference values.
     *
     * Note: PDF preferences in Readium use individual properties rather than
     * a dedicated PdfPreferences class. These values are used when creating
     * the PDF navigator.
     */
    fun toReadiumFit(): Fit? = fit?.toReadiumFit()

    fun toReadiumScroll(): Boolean = scrollMode == FlutterPdfScrollMode.VERTICAL

    fun toReadiumSpread(): Spread? = pageLayout?.toReadiumSpread()

    fun toReadiumOffsetFirstPage(): Boolean? = offsetFirstPage

    companion object {
        /**
         * Creates FlutterPdfPreferences from a JSON string.
         */
        fun fromJSON(json: String): FlutterPdfPreferences {
            return fromJSON(JSONObject(json))
        }

        /**
         * Creates FlutterPdfPreferences from a JSON object.
         */
        fun fromJSON(jsonObject: JSONObject): FlutterPdfPreferences {
            return FlutterPdfPreferences(
                fit = if (jsonObject.has("fit") && !jsonObject.isNull("fit"))
                    FlutterPdfFit.fromString(jsonObject.getString("fit"))
                else null,
                scrollMode = if (jsonObject.has("scrollMode") && !jsonObject.isNull("scrollMode"))
                    FlutterPdfScrollMode.fromString(jsonObject.getString("scrollMode"))
                else null,
                pageLayout = if (jsonObject.has("pageLayout") && !jsonObject.isNull("pageLayout"))
                    FlutterPdfPageLayout.fromString(jsonObject.getString("pageLayout"))
                else null,
                offsetFirstPage = if (jsonObject.has("offsetFirstPage") && !jsonObject.isNull("offsetFirstPage"))
                    jsonObject.getBoolean("offsetFirstPage")
                else null
            )
        }

        /**
         * Converts FlutterPdfPreferences to a JSON object.
         */
        fun toJSON(preferences: FlutterPdfPreferences): JSONObject {
            val jsonObject = JSONObject()
            preferences.fit?.let { jsonObject.put("fit", it.toFlutterString()) }
            preferences.scrollMode?.let { jsonObject.put("scrollMode", it.toFlutterString()) }
            preferences.pageLayout?.let { jsonObject.put("pageLayout", it.toFlutterString()) }
            preferences.offsetFirstPage?.let { jsonObject.put("offsetFirstPage", it) }
            return jsonObject
        }

        /**
         * Creates FlutterPdfPreferences from a Map (from Flutter method channel).
         */
        fun fromMap(prefs: Map<*, *>?): FlutterPdfPreferences {
            if (prefs == null) return FlutterPdfPreferences()

            return FlutterPdfPreferences(
                fit = (prefs["fit"] as? String)?.let { FlutterPdfFit.fromString(it) },
                scrollMode = (prefs["scrollMode"] as? String)?.let { FlutterPdfScrollMode.fromString(it) },
                pageLayout = (prefs["pageLayout"] as? String)?.let { FlutterPdfPageLayout.fromString(it) },
                offsetFirstPage = prefs["offsetFirstPage"] as? Boolean
            )
        }
    }
}

/**
 * How a PDF page fits within the viewport.
 *
 * Maps to PDFFit enum in Flutter.
 */
@Serializable
enum class FlutterPdfFit {
    WIDTH,
    CONTAIN;

    fun toReadiumFit(): Fit = when (this) {
        WIDTH -> Fit.WIDTH
        CONTAIN -> Fit.CONTAIN
    }

    fun toFlutterString(): String = when (this) {
        WIDTH -> "width"
        CONTAIN -> "contain"
    }

    companion object {
        fun fromString(value: String): FlutterPdfFit = when (value.lowercase()) {
            "width" -> WIDTH
            "contain" -> CONTAIN
            else -> CONTAIN
        }
    }
}

/**
 * Scroll direction for PDF navigation.
 *
 * Maps to PDFScrollMode enum in Flutter.
 */
@Serializable
enum class FlutterPdfScrollMode {
    HORIZONTAL,
    VERTICAL;

    fun toFlutterString(): String = when (this) {
        HORIZONTAL -> "horizontal"
        VERTICAL -> "vertical"
    }

    companion object {
        fun fromString(value: String): FlutterPdfScrollMode = when (value.lowercase()) {
            "horizontal" -> HORIZONTAL
            "vertical" -> VERTICAL
            else -> HORIZONTAL
        }
    }
}

/**
 * Page layout mode for PDF display.
 *
 * Maps to PDFPageLayout enum in Flutter.
 */
@Serializable
enum class FlutterPdfPageLayout {
    SINGLE,
    DOUBLE,
    AUTOMATIC;

    fun toReadiumSpread(): Spread = when (this) {
        SINGLE -> Spread.NEVER
        DOUBLE -> Spread.ALWAYS
        AUTOMATIC -> Spread.AUTO
    }

    fun toFlutterString(): String = when (this) {
        SINGLE -> "single"
        DOUBLE -> "double"
        AUTOMATIC -> "automatic"
    }

    companion object {
        fun fromString(value: String): FlutterPdfPageLayout = when (value.lowercase()) {
            "single" -> SINGLE
            "double" -> DOUBLE
            "automatic" -> AUTOMATIC
            else -> SINGLE
        }
    }
}

File: flureadium/macos/Classes/flureadium/model/FlutterPdfPreferences.swift
//
//  FlutterPdfPreferences.swift
//  flureadium
//
//  PDF preferences for Flutter Readium.
//  Maps to PDFPreferences in Flutter's reader_pdf_preferences.dart.
//

import Foundation
import ReadiumNavigator
import ReadiumShared

/// How a PDF page fits within the viewport.
/// Maps to PDFFit enum in Flutter.
public enum FlutterPdfFit: String {
    case width
    case contain

    /// Converts to Readium scroll mode.
    /// - width: continuous scroll mode (scroll=true)
    /// - contain: paginated mode (scroll=false)
    func toReadiumScroll() -> Bool {
        switch self {
        case .width:
            return true
        case .contain:
            return false
        }
    }

    static func fromString(_ value: String?) -> FlutterPdfFit? {
        guard let value = value else { return nil }
        return FlutterPdfFit(rawValue: value.lowercased())
    }
}

/// Scroll direction for PDF navigation.
/// Maps to PDFScrollMode enum in Flutter.
public enum FlutterPdfScrollMode: String {
    case horizontal
    case vertical

    /// Converts to Readium scroll axis.
    func toReadiumScrollAxis() -> Axis {
        switch self {
        case .horizontal:
            return .horizontal
        case .vertical:
            return .vertical
        }
    }

    static func fromString(_ value: String?) -> FlutterPdfScrollMode? {
        guard let value = value else { return nil }
        return FlutterPdfScrollMode(rawValue: value.lowercased())
    }
}

/// Page layout mode for PDF display.
/// Maps to PDFPageLayout enum in Flutter.
public enum FlutterPdfPageLayout: String {
    case single
    case double
    case automatic

    /// Converts to Readium spread mode.
    func toReadiumSpread() -> Spread {
        switch self {
        case .single:
            return .never
        case .double:
            return .always
        case .automatic:
            return .auto
        }
    }

    static func fromString(_ value: String?) -> FlutterPdfPageLayout? {
        guard let value = value else { return nil }
        return FlutterPdfPageLayout(rawValue: value.lowercased())
    }
}

/// PDF preferences for Flutter Readium.
/// Maps to PDFPreferences class in Flutter's reader_pdf_preferences.dart.
/// Navigation UX settings (gestures, edge tap, etc.) have moved to FlutterNavigationConfig.
public struct FlutterPdfPreferences {
    /// How the page fits in the viewport.
    var fit: FlutterPdfFit?

    /// Scroll orientation.
    var scrollMode: FlutterPdfScrollMode?

    /// Page layout mode.
    var pageLayout: FlutterPdfPageLayout?

    /// Whether to offset the first page in double-page spreads (for covers).
    var offsetFirstPage: Bool?

    /// Creates FlutterPdfPreferences with default values.
    init(
        fit: FlutterPdfFit? = nil,
        scrollMode: FlutterPdfScrollMode? = nil,
        pageLayout: FlutterPdfPageLayout? = nil,
        offsetFirstPage: Bool? = nil
    ) {
        self.fit = fit
        self.scrollMode = scrollMode
        self.pageLayout = pageLayout
        self.offsetFirstPage = offsetFirstPage
    }

    /// Creates FlutterPdfPreferences from a Flutter dictionary.
    init(fromMap map: [String: Any]?) {
        guard let map = map else {
            self.init()
            return
        }

        self.init(
            fit: FlutterPdfFit.fromString(map["fit"] as? String),
            scrollMode: FlutterPdfScrollMode.fromString(map["scrollMode"] as? String),
            pageLayout: FlutterPdfPageLayout.fromString(map["pageLayout"] as? String),
            offsetFirstPage: map["offsetFirstPage"] as? Bool
        )
    }

    /// Converts to Readium PDFPreferences.
    func toReadiumPreferences() -> PDFPreferences {
        var prefs = PDFPreferences()

        // Map fit to scroll mode
        if let fit = fit {
            prefs.scroll = fit.toReadiumScroll()
        }

        // Map scrollMode to scrollAxis
        if let scrollMode = scrollMode {
            prefs.scrollAxis = scrollMode.toReadiumScrollAxis()
        }

        // Map pageLayout to spread
        if let pageLayout = pageLayout {
            prefs.spread = pageLayout.toReadiumSpread()
        }

        // Direct mapping for offsetFirstPage
        if let offsetFirstPage = offsetFirstPage {
            prefs.offsetFirstPage = offsetFirstPage
        }

        return prefs
    }

    /// Converts to a dictionary for Flutter.
    func toMap() -> [String: Any] {
        var map: [String: Any] = [:]
        if let fit = fit { map["fit"] = fit.rawValue }
        if let scrollMode = scrollMode { map["scrollMode"] = scrollMode.rawValue }
        if let pageLayout = pageLayout { map["pageLayout"] = pageLayout.rawValue }
        if let offsetFirstPage = offsetFirstPage { map["offsetFirstPage"] = offsetFirstPage }
        return map
    }
}
